src/NServiceBus.Serverless/ServerlessEndpoint.cs

```diff
@@ -56,6 +59,6 @@ public async Task Process(MessageContext message)
 
         readonly SemaphoreSlim semaphoreLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);
 
-        PipelineInvoker pipeline;
+        volatile PipelineInvoker pipeline;
```

Comment:
Pretty sure this is not needed. Reference assignment  is atomic

Reply:
I think it is needed given that we're not using `lock` statements as in traditional double-checked locking where the `lock` statement creates the memory barrier. It was hard to find a clear answer to this so I went with the safer approach.

Reply:
Volatile is not needed because the async state machine does full fencing

![image](https://user-images.githubusercontent.com/174258/62229578-41c47e80-b3c0-11e9-814f-1477b7a41fee.png)

http://www.albahari.com/threading/part4.aspx

See also

https://stackoverflow.com/questions/28871878/are-you-there-asynchronously-written-value
https://stackoverflow.com/questions/53649448/is-it-needed-to-make-fields-thread-safe-when-using-async-await